### DISCUSSION TITLE
the question about reshape

### DISCUSSION THREAD

**Discussion by modestlyh** [NVlabs/sionna#595] [Q&A] [answered]
In https://nvlabs.github.io/sionna/examples/Neural_Receiver.html:

```py
 elif self._system == "neural-receiver":
            # The neural receiver computes LLRs from the frequency domain received symbols and N0
            y = tf.squeeze(y, axis=1)
            llr = self._neural_receiver([y, no])
            llr = insert_dims(llr, 2, 1) # Reshape the input to fit what the resource grid demapper is expected
            llr = self._rg_demapper(llr) # Extract data-carrying resource elements. The other LLrs are discarded
            llr = tf.reshape(llr, [batch_size, 1, 1, n]) # Reshape the LLRs to fit what the outer decoder is expected
    # Outer coding is not needed if the information rate is returned
        if self._training:
            # Compute and return BMD rate (in bit), which is known to be an achievable
            # information rate for BICM systems.
            # Training aims at maximizing the BMD rate
            bce = tf.nn.sigmoid_cross_entropy_with_logits(c, llr)
            bce = tf.reduce_mean(bce)
            rate = tf.constant(1.0, tf.float32) - bce/tf.math.log(2.)
            return rate

 llr = self._rg_demapper(llr)
```

the rg_demapper needs the shape of input :`[batch_size, num_rx, num_streams_per_rx, num_ofdm_symbols, fft_size, data_dim]`,if num_rx and num_streams_per_rx do not equal to 1, we will need reshape ,but I then meet a  question:
 Input to reshape is a tensor with XXX values, but the requested shape has XXX
I set  num_ut = 4, num_ut_ant=1,num_bs=1,num_bs=16, so I meet the question

**Comment by jhoydis** [comment]
Hi @modestlyh,

Could you please try to rephrase your question? I am not sure that I fully understand your problem.

**Reply by modestlyh** [reply]
Okay, Neural Receiver for OFDM SIMO Systems is provided in Sionna's documentation, in this documentation, the number of users is 1, the number of user antennas is 1, the number of base stations is 1, the number of base station antennas is 2. I would like to change the number of user antennas to 4, and the number of base station antennas to 16 in order to change it to a MIMO system, the following code needs to be changed as well:

```py
elif self._system == "neural-receiver":
            # The neural receiver computes LLRs from the frequency domain received symbols and N0
            y = tf.squeeze(y, axis=1)
            llr = self._neural_receiver([y, no]) #llr.shape [batch size, num ofdm symbols, num subcarriers, num_bits_per_symbol]
            llr = insert_dims(llr, 2, 1) # Reshape the input to fit what the resource grid demapper is expected
           # llr.shape [batch size, 1, 1, num ofdm symbols, num subcarriers, num_bits_per_symbol]
            llr = self._rg_demapper(llr) # Extract data-carrying resource elements. The other LLrs are discarded
            llr = tf.reshape(llr, [batch_size, 1, 1, n]) # Reshape the LLRs to fit what the outer decoder is expected
         # [batch_size, 1, 1, n] is c(codeword).shape , The first 1 represents the number of users and the second 1 represents the number of user antennas.
```

this is the input of _rg_demapper: [batch_size, num_rx, num_streams_per_rx, num_ofdm_symbols, fft_size, data_dim]
num_streams_per_rx  is int(self.num_tx*self.num_streams_per_tx/self.num_rx)
self.num_tx=1,self.num_streams_per_tx=num_tx_ant=4,self.num_rx=1
so num_streams_per_rx=4
llr needs to reshape  [batch size, 1, 4, num ofdm symbols, num subcarriers, num_bits_per_symbol]  instead of  [batch size, 1, 1, num ofdm symbols, num subcarriers, num_bits_per_symbol]  before    llr = self._rg_demapper(llr)
And  The last line of code needs to be changed to    llr = tf.reshape(llr, [batch_size, 1, 4, n]) 
But this will encounter a problem that the total number of values does not match，because reshape needs to have the same total number of values
What should I do?

**Comment by jhoydis** [comment, marked as answer]
Im short, the neural receiver architecture is for a SIMO system, i.e., a single stream is sent. You cannot have more than one stream per receiver.

Please have a look our [recent repo](https://github.com/NVlabs/neural_rx) in which we extend the neural receiver architecture to MIMO systems with multiple users.

**Reply by modestlyh** [reply]
extremely grateful！！！
